cli/src/cmd/forge/script/receipts.rs

```diff
@@ -1,49 +1,45 @@
-use std::collections::BTreeMap;
-
-use crate::{
-    cmd::ScriptSequence,
-    utils::{get_http_provider, print_receipt},
-};
-use ethers::{
-    abi::Address,
-    prelude::{Bytes, Middleware, TransactionReceipt, TxHash, U256},
-    types::transaction::eip2718::TypedTransaction,
-};
+use crate::{cmd::ScriptSequence, utils::print_receipt};
+use ethers::prelude::{Http, Middleware, Provider, TransactionReceipt, H256, U256};
 use futures::future::join_all;
 
 use super::broadcast::BroadcastError;
 
-// Get all transactions in the `tx_hashes` list.
-pub async fn get_pending_txes(
-    tx_hashes: &Vec<TxHash>,
-    fork_url: &str,
-) -> BTreeMap<(Address, U256), (Bytes, TxHash)> {
-    let provider = get_http_provider(fork_url);
-    let mut pending_txes = BTreeMap::new();
-    for pending in tx_hashes {
-        if let Ok(Some(tx)) = provider.get_transaction(*pending).await {
-            pending_txes.insert((tx.from, tx.nonce), (tx.input, *pending));
+/// Gets the receipts of previously pending transactions.
+pub async fn wait_for_pending(
+    provider: &Provider<Http>,
+    deployment_sequence: &mut ScriptSequence,
+) -> eyre::Result<()> {
+    if !deployment_sequence.pending.is_empty() {
+        println!("##\nChecking previously pending transactions.");
+        let mut future_receipts = vec![];
+
+        for tx_hash in &deployment_sequence.pending {
+            future_receipts.push(pending_receipt(provider, *tx_hash));
         }
+
+        wait_for_receipts(future_receipts, deployment_sequence).await?;
     }
-    pending_txes
+    Ok(())
 }
 
-/// Given a tx, it checks if there is already a receipt for it. Compares `from`, `nonce` and `data`.
-pub async fn maybe_has_receipt(
-    tx: &TypedTransaction,
-    pending_txes: &BTreeMap<(Address, U256), (Bytes, TxHash)>,
-    fork_url: &str,
-) -> Option<TransactionReceipt> {
-    let mut receipt = None;
-    if let Some((data, tx_hash)) = pending_txes.get(&(*tx.from().unwrap(), *tx.nonce().unwrap())) {
-        if tx.data().unwrap().eq(data) {
-            let provider = get_http_provider(fork_url);
-            if let Ok(ret) = provider.get_transaction_receipt(*tx_hash).await {
-                receipt = ret
-            }
-        }
-    }
-    receipt
+/// Waits for a pending receipt, and gets its nonce to return (receipt, nonce).
+async fn pending_receipt(
+    provider: &Provider<Http>,
+    tx_hash: H256,
+) -> Result<(TransactionReceipt, U256), BroadcastError> {
+    let pending_err =
+        BroadcastError::Simple(format!("Failed to get pending transaction {tx_hash}"));
```

Comment:
you can probably turn this into a closure that creates the error and get rid of the clones